In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

In [2]:
#read in our data
df_full = pd.read_csv('data/2018_missoula_marathon_full.csv')
df_half = pd.read_csv('data/2018_missoula_marathon_half.csv')

### Cleaning & Wrangling

In [3]:
df_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1034 entries, 0 to 1033
Data columns (total 12 columns):
Place        1034 non-null int64
Name         1034 non-null object
City         1012 non-null object
Bib_No       1034 non-null int64
Age          1034 non-null int64
Gender       1034 non-null object
Age_Group    1034 non-null object
Chip_Time    1034 non-null object
Gun_Time     1034 non-null object
Chip_Diff    1028 non-null object
Pace         1034 non-null object
Race         1034 non-null object
dtypes: int64(3), object(9)
memory usage: 97.0+ KB


In [4]:
df_half.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2659 entries, 0 to 2658
Data columns (total 12 columns):
Place        2659 non-null int64
Name         2659 non-null object
City         2645 non-null object
Bib_No       2659 non-null int64
Age          2658 non-null float64
Gender       2659 non-null object
Age_Group    2659 non-null object
Chip_Time    2659 non-null object
Gun_Time     2659 non-null object
Chip_Diff    2647 non-null object
Pace         2659 non-null object
Race         2659 non-null object
dtypes: float64(1), int64(2), object(9)
memory usage: 249.4+ KB


Looking at the info, it appears that we have issues with missing 'City' fields and missingt 'Chip_Diff' fields.

All missing 'City' entries are for race pacers. Let's make an assumption that they are all from Missoula. The missing 'Chip_Diff' entries occur when the chip time and the gun time are the same, so let's replace those with zeros.

There's one more issue to deal with. We're missing an age for one male half marathon runner (bib number 9384). That runner is listed as 'Unknown Participant' and is missing other details as well. Let's drop that record.

In [5]:
#first let's drop our rogue runner
df_half = df_half[df_half.Bib_No != 9384]
df_half.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2658 entries, 0 to 2658
Data columns (total 12 columns):
Place        2658 non-null int64
Name         2658 non-null object
City         2645 non-null object
Bib_No       2658 non-null int64
Age          2658 non-null float64
Gender       2658 non-null object
Age_Group    2658 non-null object
Chip_Time    2658 non-null object
Gun_Time     2658 non-null object
Chip_Diff    2646 non-null object
Pace         2658 non-null object
Race         2658 non-null object
dtypes: float64(1), int64(2), object(9)
memory usage: 270.0+ KB


In [6]:
df_half[df_half.City.isna()]

Place                        Name City  Bib_No   Age Gender  \
87       88        Kate Johnson - Pacer  NaN    3400  26.0      F   
519     520     Ilana Abrahamson -Pacer  NaN    2107  42.0      F   
520     521     Sara McAllister - Pacer  NaN    3819  35.0      F   
630     631         Rachel Toor - Pacer  NaN    7029  56.0      F   
943     944     Theneshia Glaze - Pacer  NaN    3011  41.0      F   
978     979       Rebecca Garner -Pacer  NaN    2971  52.0      F   
1903     51          Dillon May - Pacer  NaN    3809  21.0      M   
1904     52     Nicolas Composto -Pacer  NaN    2578  25.0      M   
1997    145         Scott Davis - Pacer  NaN    2689  26.0      M   
2184    332     Danny Tenenbaum - Pacer  NaN    4799  32.0      M   
2185    333  David Schmetterling -Pacer  NaN    4529  46.0      M   
2319    467       Vo Von Sehlen - Pacer  NaN    7031  53.0      M   
2341    489      Keith Fichtner - Pacer  NaN    2886  42.0      M   

          Age_Group    Chip_Time    Gun_Time Chip_Diff     Pace  Race  
87     14/201:25-29   1:45:16.99  1:45:26.48   0:09.49   8:02/M  half  
519    68/211:40-44   2:09:46.25  2:10:25.72   0:39.47   9:54/M  half  
520   103/293:35-39   2:09:46.70  2:10:26.13   0:39.43   9:54/M  half  
630    24/152:55-59   2:14:36.70  2:15:48.73   1:12.03  10:16/M  half  
943   117/211:40-44   2:29:50.89  2:31:52.41   2:01.52  11:26/M  half  
978    68/173:50-54   2:31:05.50  2:33:06.82   2:01.32  11:32/M  half  
1903    12/32:20-24   1:29:38.56  1:29:42.16   0:03.60   6:51/M  half  
1904     5/68:25-29   1:29:38.79  1:29:42.21   0:03.42   6:51/M  half  
1997    15/68:25-29   1:45:17.19  1:45:26.60   0:09.41   8:02/M  half  
2184    56/97:30-34   2:00:16.96  2:00:39.64   0:22.68   9:11/M  half  
2185    34/91:45-49   2:00:17.55  2:00:40.14   0:22.59   9:11/M  half  
2319    39/69:50-54   2:14:34.52  2:15:49.42   1:14.90  10:16/M  half  
2341    53/84:40-44   2:17:38.43  2:19:03.49   1:25.06  10:30/M  half

In [7]:
df_half.City = df_half.City.fillna('Missoula MT US')

In [8]:
df_half[df_half.City.isna()]

Empty DataFrame
Columns: [Place, Name, City, Bib_No, Age, Gender, Age_Group, Chip_Time, Gun_Time, Chip_Diff, Pace, Race]
Index: []

In [9]:
df_half[df_half.Chip_Diff.isna()]

Place               Name                   City  Bib_No   Age Gender  \
707     708     Emily Cheroske        Missoula  MT US    5182  22.0      F   
1007   1008     Claire Rappold        Missoula  MT US    4333  25.0      F   
1012   1013      Janice Loudon   Overland Park  KS US    3703  58.0      F   
1659   1660    Marybeth Healey   Coeur D Alene  ID US    3160  57.0      F   
1788   1789         Dawn Clark           Butte  MT US    2546  59.0      F   
1808   1809      Amanda Ledoux  Spokane Valley  WA US    3644  44.0      F   
1809   1810      Celina Taylor         Spokane  WA US    4788  42.0      F   
1844   1845        Alina McCue        Missoula  MT US    5185  39.0      F   
1889     37      Connor Morris    Port Orchard  WA US    5298  20.0      M   
1923     71  Lucca Wellenstein        Missoula  MT US    4993  16.0      M   
1974    122         Ian Curtis        Missoula  MT US    2649  16.0      M   
2110    258       Tom Frandsen            Lolo  MT US    2930  34.0      M   

          Age_Group    Chip_Time    Gun_Time Chip_Diff     Pace  Race  
707    66/121:20-24   2:18:57.49  2:18:57.49       NaN  10:36/M  half  
1007  121/201:25-29   2:33:05.99  2:33:05.99       NaN  11:41/M  half  
1012   62/152:55-59   2:33:41.21  2:33:41.21       NaN  11:44/M  half  
1659  128/152:55-59   3:32:18.86  3:32:18.86       NaN  16:12/M  half  
1788  146/152:55-59   4:00:36.85  4:00:36.85       NaN  18:22/M  half  
1808  202/211:40-44   4:09:38.16  4:09:38.16       NaN  19:03/M  half  
1809  203/211:40-44   4:09:38.62  4:09:38.62       NaN  19:03/M  half  
1844  293/293:35-39   4:31:55.15  4:31:55.15       NaN  20:45/M  half  
1889    11/32:20-24   1:26:38.11  1:26:38.11       NaN   6:37/M  half  
1923    16/45:15-19   1:33:38.00  1:33:38.00       NaN   7:09/M  half  
1974    25/45:15-19   1:41:59.79  1:41:59.79       NaN   7:47/M  half  
2110    40/97:30-34   1:55:03.88  1:55:03.88       NaN   8:47/M  half

In [10]:
df_half.Chip_Diff = df_half.Chip_Diff.fillna(0)
df_half[df_half.Chip_Diff.isna()]

Empty DataFrame
Columns: [Place, Name, City, Bib_No, Age, Gender, Age_Group, Chip_Time, Gun_Time, Chip_Diff, Pace, Race]
Index: []

In [11]:
#let's fix those same issues with the full results.
df_full.City = df_full.City.fillna('Missoula MT US')
df_full.Chip_Diff = df_full.Chip_Diff.fillna(0)

In [12]:
df_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1034 entries, 0 to 1033
Data columns (total 12 columns):
Place        1034 non-null int64
Name         1034 non-null object
City         1034 non-null object
Bib_No       1034 non-null int64
Age          1034 non-null int64
Gender       1034 non-null object
Age_Group    1034 non-null object
Chip_Time    1034 non-null object
Gun_Time     1034 non-null object
Chip_Diff    1034 non-null object
Pace         1034 non-null object
Race         1034 non-null object
dtypes: int64(3), object(9)
memory usage: 97.0+ KB


In [13]:
df_half.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2658 entries, 0 to 2658
Data columns (total 12 columns):
Place        2658 non-null int64
Name         2658 non-null object
City         2658 non-null object
Bib_No       2658 non-null int64
Age          2658 non-null float64
Gender       2658 non-null object
Age_Group    2658 non-null object
Chip_Time    2658 non-null object
Gun_Time     2658 non-null object
Chip_Diff    2658 non-null object
Pace         2658 non-null object
Race         2658 non-null object
dtypes: float64(1), int64(2), object(9)
memory usage: 270.0+ KB


Ok. We have our missing/messy values worked out. Next, work our our columns.

The 'Age_Group' field is holding both the runners place out of the total for that age group and the label. For example:

66/121:20-24 - The runner was number 66 out of 121 runners in the 20-24 age group.

Let's drop the total number in the group. We can always calculate that if we need it. And let's split the data into two colums one for place and one for group.

In [14]:
df_half['Age_Group_Place'], df_half['Age_Group'] = df_half.Age_Group.str.split(':', 1).str
df_half['Age_Group_Place'], _ = df_half.Age_Group_Place.str.split('/', 1).str

In [15]:
df_full['Age_Group_Place'], df_full['Age_Group'] = df_full.Age_Group.str.split(':', 1).str
df_full['Age_Group_Place'], _ = df_full.Age_Group_Place.str.split('/', 1).str

Let's trim off the '/M' in Pace in case we want to use it numerically.

In [16]:
df_full.Pace = df_full.Pace.str[:-2]
df_half.Pace = df_half.Pace.str[:-2]

In [17]:
df_half.head()

Place                 Name                     City  Bib_No   Age Gender  \
0      1        Makena Morley           Bigfork  MT US    2051  21.0      F   
1      2     Misiker Demessie  Colorado Springs  CO US    2060  31.0      F   
2      3         Kelsi Lasota        Estes Park  CO US    2052  22.0      F   
3      4  Elizabeth Wasserman            Goshen  NY US    2056  23.0      F   
4      5     Andrea Masterson           Seattle  WA US    2053  21.0      F   

  Age_Group    Chip_Time    Gun_Time Chip_Diff  Pace  Race Age_Group_Place  
0     20-24   1:15:50.45  1:15:51.79   0:01.34  5:47  half               1  
1     30-34   1:18:03.69  1:18:05.58   0:01.89  5:57  half               1  
2     20-24   1:18:41.68  1:18:43.68   0:02.00  6:00  half               2  
3     20-24   1:19:18.71  1:19:20.92   0:02.21  6:03  half               3  
4     20-24   1:21:38.93  1:21:40.32   0:01.39  6:14  half               4

In [18]:
df_full.head()

Place            Name                City  Bib_No  Age Gender Age_Group  \
0      1     Keely Baker  Great Falls  MT US      52   26      F     25-29   
1      2  Trisha Drobeck     Missoula  MT US      51   38      F     35-39   
2      3    Chelsea Dana     Billings  MT US     355   35      F     35-39   
3      4       Evie Tate     Missoula  MT US    1211   23      F     20-24   
4      5      Erin Forde     Missoula  MT US      56   25      F     25-29   

     Chip_Time    Gun_Time Chip_Diff  Pace  Race Age_Group_Place  
0   2:51:46.12  2:51:48.01   0:01.89  6:33  full               1  
1   2:53:28.28  2:53:30.08   0:01.80  6:37  full               1  
2   3:01:37.43  3:01:43.04   0:05.61  6:56  full               2  
3   3:09:32.24  3:09:45.38   0:13.14  7:14  full               1  
4   3:09:37.84  3:09:41.66   0:03.82  7:14  full               2

Looks good. Let's combine and save the output...in hindsight, we should have combined sooner and saved ourselves some steps, but I didn't think I was going to until just now. Oh well, live and learn.

In [19]:
df_all = pd.concat([df_full, df_half])

In [20]:
df_all.sample(10)

Place                Name                   City  Bib_No   Age Gender  \
114     115          Lisa Eiler        Missoula  MT US    5181  40.0      F   
516      44      Terence Palmer        Edmonton  AB CA     929  40.0      M   
43       44       Kailee Carnes        Missoula  MT US    7015  30.0      F   
1412   1413  Samantha Tartaglia      Hope Mills  NC US    4784  27.0      F   
502      30      Shea Gingerich       Corvallis  MT US    5313  27.0      M   
102     103        Cyndi Rickey    Lake Jackson  TX US    1021  56.0      F   
668     669        Sarah Stover  Thompson Falls  MT US    4746  41.0      F   
1516   1517        Dulce Barton      Long Beach  CA US    2246  62.0      F   
255     256    Stephanie Marron       Vancouver  BC CA     772  35.0      F   
452     453       Andrea Haines        Missoula  MT US     509  37.0      F   

     Age_Group    Chip_Time    Gun_Time Chip_Diff   Pace  Race Age_Group_Place  
114      40-44   1:48:33.77  1:48:50.97   0:17.20   8:17  half              13  
516      40-44   3:15:54.54  3:16:02.21   0:07.67   7:29  full               9  
43       30-34   1:40:11.21  1:40:19.10   0:07.89   7:39  half              10  
1412     25-29   3:01:52.12  3:04:17.50   2:25.38  13:53  half             162  
502      25-29   3:03:32.47  3:03:38.44   0:05.97   7:00  full              10  
102      55-59   4:04:17.90  4:04:45.29   0:27.39   9:19  full               3  
668      40-44   2:17:00.90  2:18:31.27   1:30.37  10:27  half              84  
1516     60-64   3:12:24.71  3:14:30.07   2:05.36  14:41  half              69  
255      35-39   4:52:51.31  4:53:40.68   0:49.37  11:11  full              41  
452      35-39   6:44:26.77  6:46:13.28   1:46.51  15:26  full              70

Let's save it and move on to analysis.

In [21]:
df_all['Chip_Time_Minutes'] = pd.to_timedelta(df_all.Chip_Time).dt.total_seconds()/60
df_all['Gun_Time_Minutes'] = pd.to_timedelta(df_all.Gun_Time).dt.total_seconds()/60
df_all['Chip_Diff_Seconds'] = (df_all.Gun_Time_Minutes - df_all.Chip_Time_Minutes)*60

In [22]:
df_all.head()

Place            Name                City  Bib_No   Age Gender Age_Group  \
0      1     Keely Baker  Great Falls  MT US      52  26.0      F     25-29   
1      2  Trisha Drobeck     Missoula  MT US      51  38.0      F     35-39   
2      3    Chelsea Dana     Billings  MT US     355  35.0      F     35-39   
3      4       Evie Tate     Missoula  MT US    1211  23.0      F     20-24   
4      5      Erin Forde     Missoula  MT US      56  25.0      F     25-29   

     Chip_Time    Gun_Time Chip_Diff  Pace  Race Age_Group_Place  \
0   2:51:46.12  2:51:48.01   0:01.89  6:33  full               1   
1   2:53:28.28  2:53:30.08   0:01.80  6:37  full               1   
2   3:01:37.43  3:01:43.04   0:05.61  6:56  full               2   
3   3:09:32.24  3:09:45.38   0:13.14  7:14  full               1   
4   3:09:37.84  3:09:41.66   0:03.82  7:14  full               2   

   Chip_Time_Minutes  Gun_Time_Minutes  Chip_Diff_Seconds  
0         171.768667        171.800167               1.89  
1         173.471333        173.501333               1.80  
2         181.623833        181.717333               5.61  
3         189.537333        189.756333              13.14  
4         189.630667        189.694333               3.82

In [23]:
df_all['Pace_Decimal'] = np.where(df_all.Race == 'full', df_all.Chip_Time_Minutes/26.2, df_all.Chip_Time_Minutes/13.1)

Fix our data types

In [24]:
df_all.Race = df_all.Race.astype('category')
df_all.Age_Group_Place = df_all.Age_Group_Place.astype('int32')

In [25]:
df_all.describe()

Place       Bib_No          Age  Age_Group_Place  \
count  3692.000000  3692.000000  3692.000000      3692.000000   
mean    626.201788  2957.951246    41.475623        65.133261   
std     505.587348  1632.593848    13.657386        61.065528   
min       1.000000     1.000000    10.000000         1.000000   
25%     231.000000  1298.750000    31.000000        20.000000   
50%     462.000000  3136.500000    40.000000        45.000000   
75%     930.250000  4313.250000    52.000000        89.000000   
max    1853.000000  7032.000000    85.000000       293.000000   

       Chip_Time_Minutes  Gun_Time_Minutes  Chip_Diff_Seconds  Pace_Decimal  
count        3692.000000       3692.000000        3692.000000   3692.000000  
mean          186.360980        188.007287          98.778448     11.252631  
std            73.533441         73.693673          80.275205      2.891586  
min            67.856500         67.856500           0.000000      5.179885  
25%           129.073458        130.385250          32.625000      9.131054  
50%           163.896333        166.416917          78.815000     10.753575  
75%           231.915833        233.535417         145.535000     12.792576  
max           472.856333        473.926167         737.910000     30.162748

In [26]:
#I think we're done with our cleaning and feature engineering. Let's save a copy.
df_all.to_csv('data/2018_missoula_marathon_all_clean.csv', index=False)

### Exploration

Let's start by looking at some of our numeric values.

In [27]:
#let's start by updating our 'half' and 'full' data frames.
df_full = df_all[df_all.Race == 'full']
df_half = df_all[df_all.Race == 'half']

In [28]:
df_full.describe()

Place       Bib_No          Age  Age_Group_Place  \
count  1034.000000  1034.000000  1034.000000      1034.000000   
mean    260.872340   833.135397    41.327853        30.250484   
std     152.515542   763.499283    12.466329        20.927342   
min       1.000000     1.000000    13.000000         1.000000   
25%     130.000000   430.250000    31.000000        12.250000   
50%     259.000000   773.500000    40.000000        27.000000   
75%     388.000000  1113.750000    51.000000        45.000000   
max     561.000000  7012.000000    80.000000        89.000000   

       Chip_Time_Minutes  Gun_Time_Minutes  Chip_Diff_Seconds  Pace_Decimal  
count        1034.000000       1034.000000        1034.000000   1034.000000  
mean          278.160529        279.054454          53.635493     10.616814  
std            58.572159         59.052299          36.166043      2.235579  
min           147.876500        147.891667           0.000000      5.644141  
25%           235.323250        235.769458          22.160000      8.981803  
50%           270.533500        271.499583          48.320000     10.325706  
75%           314.664333        315.965833          82.325000     12.010089  
max           472.856333        473.926167         178.840000     18.047952

### Full
**Age**
- Min/Max: 13/80
- Mean: 41.33
- Std Dev: 12.47

**Chip Time in Minutes**
- Min/Max: 147.88/472.86
- Mean: 278.16
- Std Dev: 59.05

**Pace**
- Min/Max: 5.64/18.05
- Mean: 10.62
- Std Dev: 2.24

In [29]:
df_half.describe()

Place       Bib_No          Age  Age_Group_Place  \
count  2658.000000  2658.000000  2658.000000       2658.00000   
mean    768.320166  3784.534989    41.533108         78.70316   
std     523.361209  1017.588079    14.095374         65.97093   
min       1.000000   732.000000    10.000000          1.00000   
25%     333.000000  2920.250000    31.000000         26.00000   
50%     665.000000  3772.500000    40.000000         60.00000   
75%    1188.750000  4635.750000    52.000000        117.00000   
max    1853.000000  7032.000000    85.000000        293.00000   

       Chip_Time_Minutes  Gun_Time_Minutes  Chip_Diff_Seconds  Pace_Decimal  
count        2658.000000       2658.000000        2658.000000   2658.000000  
mean          150.649642        152.588638         116.339703     11.499973  
std            40.278087         41.321930          85.684305      3.074663  
min            67.856500         67.856500           0.000000      5.179885  
25%           120.931167        121.994625          40.975000      9.231387  
50%           143.679667        145.735917         102.980000     10.967913  
75%           172.899375        175.774125         182.555000     13.198426  
max           395.132000        400.393500         737.910000     30.162748

### Half
**Age**
- Min/Max: 10/85
- Mean: 41.53
- Std Dev: 14.1

**Chip Time in Minutes**
- Min/Max: 67.86/395.13
- Mean: 150.65
- Std Dev: 40.28

**Pace**
- Min/Max: 5.18/30.16
- Mean: 11.5
- Std Dev: 3.07